File: Source/M2Engine/FilterTexts.pas
unit FilterTexts;

interface

uses
  Windows, Classes, SysUtils;

type
  TFilterMsg = record
    Msg: string;
    Replace: string;
  end;

  pTFilterMsg = ^TFilterMsg;

  TFilterTexts = class
  private
    FRecordCount: Integer;
    FList: TList;
    function TrimAll(Text: string): string;
    function GetCount: Integer;
    function GetItems(Index: Integer): pTFilterMsg;
  public
    constructor Create();
    destructor Destroy; override;
    procedure LoadFromFile;
    procedure SaveToFile;
    function Filter(sMsg: string; var sNewMsg: string): Boolean;
    function Find(sMsg: string): Boolean;
    function Add(sMsg, sReplaceMsg: string): Boolean;
    function Delete(sMsg: string): Boolean;
    property Items[Index: Integer]: pTFilterMsg read GetItems;
    property Count: Integer read GetCount;
    property RecordCount: Integer read FRecordCount;
  end;

implementation

uses
  StrUtils, M2Share, HUtil32;

const
  TextChars = [#32 .. #255];

function GetWebSubAddr(sDomainName: string; var sName: string): string;
begin
  Result := sDomainName;
  // sAddr := GetWebAddr(sAddr);
  if Pos('.', sDomainName) > 0 then
    Result := Copy(sDomainName, Pos('.', sDomainName) + 1, Length(sDomainName));
  sName := Copy(Result, 1, Pos('.', Result) - 1);
  // Result := sAddr;
end;

constructor TFilterTexts.Create();
begin
  FList := TList.Create;
  FRecordCount := 0;
end;

destructor TFilterTexts.Destroy;
var
  I: Integer;
begin
  for I := 0 to FList.Count - 1 do
  begin
    Dispose(pTFilterMsg(FList.Items[I]));
  end;
  FList.Free;
  inherited;
end;

function TFilterTexts.TrimAll(Text: string): string;
var
  I: Integer;
begin
  Text := Trim(Text);
  for I := Length(Text) downto 1 do
  begin
{$IF CompilerVersion >= 22}
    if (not CharInSet(Text[I], TextChars)) then
      SysTem.Delete(Text, I, 1);
{$ELSE}
    if (not(Text[I] in TextChars)) then
      SysTem.Delete(Text, I, 1);
{$IFEND}
  end;
  Result := Text;
end;

function TFilterTexts.GetCount: Integer;
begin
  Result := FList.Count;
end;

function TFilterTexts.GetItems(Index: Integer): pTFilterMsg;
begin
  Result := FList.Items[Index];
end;

function TFilterTexts.Find(sMsg: string): Boolean;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
begin
  Result := False;
  sMsg := TrimAll(sMsg);
  for I := 0 to FList.Count - 1 do
  begin
    FilterMsg := pTFilterMsg(FList.Items[I]);
    if (CompareText(sMsg, FilterMsg.Msg) = 0) then
    begin
      Result := True;
      Break;
    end;
  end;
end;

function TFilterTexts.Filter(sMsg: string; var sNewMsg: string): Boolean;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
  IsSameText: Boolean;
begin
  Result := False;
  sNewMsg := sMsg;
  // try
  sNewMsg := TrimAll(sNewMsg);
  if CanFilterMsg(sNewMsg) then
  begin
    if (CompareText(sNewMsg, '$') = 0) or AnsiContainsText(sNewMsg, '$') then
    begin
      Result := True;
      if (Length(sNewMsg) = 1) then
      begin
        sNewMsg := '';
      end
      else
      begin
        sNewMsg := AnsiReplaceText(sNewMsg, '$', '');
      end;
      Exit;
    end;

    for I := 0 to FList.Count - 1 do
    begin
      FilterMsg := pTFilterMsg(FList.Items[I]);
      IsSameText := SameText(sNewMsg, FilterMsg.Msg);
      if IsSameText or AnsiContainsText(sNewMsg, FilterMsg.Msg) then
      begin
        Result := True;
        if FilterMsg.Replace = '' then
        begin
          sNewMsg := '';
        end
        else
        begin
          if IsSameText then
          begin
            sNewMsg := FilterMsg.Replace;
          end
          else
          begin
            sNewMsg := AnsiReplaceText(sNewMsg, FilterMsg.Msg, FilterMsg.Replace);
          end;
        end;
        if sNewMsg = '' then
          Break;
        // Break;
      end;
    end;
  end;
  // except
  // MainOutMessage('TFilterTexts.Filter:'+IntToStr(nError));
  // end;
end;

function TFilterTexts.Add(sMsg, sReplaceMsg: string): Boolean;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
begin
  Result := False;
  sMsg := TrimAll(sMsg);
  for I := 0 to FList.Count - 1 do
  begin
    FilterMsg := pTFilterMsg(FList.Items[I]);
    if (CompareText(sMsg, FilterMsg.Msg) = 0) then
    begin
      Exit;
    end;
  end;
  if CanFilterMsg(sMsg) then
  begin
    New(FilterMsg);
    FilterMsg.Msg := sMsg;
    FilterMsg.Replace := sReplaceMsg;
    FList.Add(FilterMsg);
    Inc(FRecordCount);
    Result := True;
  end;
end;

function TFilterTexts.Delete(sMsg: string): Boolean;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
begin
  Result := False;
  for I := 0 to FList.Count - 1 do
  begin
    FilterMsg := pTFilterMsg(FList.Items[I]);
    if (CompareText(sMsg, FilterMsg.Msg) = 0) then
    begin
      Dec(FRecordCount);
      FList.Delete(I);
      Dispose(FilterMsg);
      Result := True;
      Break;
    end;
  end;
end;

procedure TFilterTexts.LoadFromFile;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
  LoadList: TStringList;
  sFileName: string;
  sLineText: string;
  sMsg: string;
  sReplace: string;
begin
  sFileName := g_Config.sEnvirDir + 'FilterMsgList.txt';
  if FileExists(sFileName) then
  begin
    LoadList := TStringList.Create;
    LoadList.LoadFromFile(sFileName);
    for I := 0 to LoadList.Count - 1 do
    begin
      sLineText := LoadList.Strings[I];
      if (sLineText <> '') and (sLineText[1] <> ';') then
      begin
        sLineText := GetValidStr3(sLineText, sMsg, [#9]);
        sLineText := GetValidStr3(sLineText, sReplace, [#9]);
        if CanFilterMsg(sMsg) then
        begin
          New(FilterMsg);
          FilterMsg.Msg := sMsg;
          FilterMsg.Replace := sReplace;
          FList.Add(FilterMsg);
          Inc(FRecordCount);
        end;
      end;
    end;
    LoadList.Free;
  end;
end;

procedure TFilterTexts.SaveToFile;
var
  I: Integer;
  FilterMsg: pTFilterMsg;
  SaveList: TStringList;
  sFileName: string;
begin
  SaveList := TStringList.Create;
  for I := 0 to FList.Count - 1 do
  begin
    FilterMsg := pTFilterMsg(FList.Items[I]);
    SaveList.Add(FilterMsg.Msg + #9 + FilterMsg.Replace);
  end;
  sFileName := g_Config.sEnvirDir + 'FilterMsgList.txt';
  try
    SaveList.SaveToFile(sFileName);
  except

  end;
  SaveList.Free;
end;

end.

File: GXX.CSharp/src/GXX.M2Server/Engine/FilterTexts.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using GXX.Core.Util;

namespace GXX.M2Server.Engine;

/// <summary>FilterTexts.pas TFilterMsg record 镜像。</summary>
public sealed class TFilterMsg
{
    public string Msg = "";
    public string Replace = "";
}

/// <summary>
/// FilterTexts.pas TFilterTexts 1:1（批次J56）：TrimAll（删除 #32..#255 之外字符）、
/// Find/Add/Delete 全部 CompareText（不区分大小写）、Add 重名拒绝（CanFilterMsg 恒真）、
/// Filter 替换核心（'$' 特例：包含 → 全滤且 len=1 置空串否则删 '$'；SameText 全等 → Replace（空则清空）；
/// AnsiContainsText 包含 → 大小写不敏感替换；结果空串即停）、LoadFromFile/SaveToFile（FilterMsgList.txt
/// TAB 分隔、';' 注释跳过）。g_FilterTexts 单元全局由窗体层持有。
/// </summary>
public class TFilterTexts
{
    private readonly List<TFilterMsg> _list = new();
    private int _recordCount;

    public int Count => _list.Count;

    public int RecordCount => _recordCount;

    public TFilterMsg GetItems(int index) => _list[index];

    /// <summary>
    /// TrimAll：Trim 后删除控制字符。Delphi TextChars=[#32..#255] 基于 ANSI 字节流
    /// （GBK 汉字字节 128..255 全部保留），Unicode 下的 1:1 等价 = 仅删除 &lt;#32。
    /// </summary>
    public static string TrimAll(string text)
    {
        text = text.Trim();
        var sb = new StringBuilder(text.Length);
        foreach (var ch in text)
        {
            if (ch >= (char)32)
                sb.Append(ch);
        }
        return sb.ToString();
    }

    public bool Find(string sMsg)
    {
        sMsg = TrimAll(sMsg);
        foreach (var filterMsg in _list)
        {
            if (string.Compare(sMsg, filterMsg.Msg, StringComparison.OrdinalIgnoreCase) == 0)
                return true;
        }
        return false;
    }

    /// <summary>Filter（119-176）：替换命中返回 true 且 sNewMsg 为替换产物。</summary>
    public bool Filter(string sMsg, out string sNewMsg)
    {
        bool result = false;
        sNewMsg = TrimAll(sMsg);
        if (!M2ShareFuncs.CanFilterMsg(sNewMsg))
            return result;

        // '$' 特例
        if (string.Compare(sNewMsg, "$", StringComparison.OrdinalIgnoreCase) == 0 ||
            sNewMsg.IndexOf('$', StringComparison.OrdinalIgnoreCase) >= 0)
        {
            result = true;
            sNewMsg = sNewMsg.Length == 1 ? "" : sNewMsg.Replace("$", "");
            return result;
        }

        foreach (var filterMsg in _list)
        {
            bool isSameText = string.Equals(sNewMsg, filterMsg.Msg, StringComparison.OrdinalIgnoreCase);
            if (isSameText || sNewMsg.IndexOf(filterMsg.Msg, StringComparison.OrdinalIgnoreCase) >= 0)
            {
                result = true;
                if (filterMsg.Replace.Length == 0)
                {
                    sNewMsg = "";
                }
                else
                {
                    sNewMsg = isSameText
                        ? filterMsg.Replace
                        : AnsiReplaceText(sNewMsg, filterMsg.Msg, filterMsg.Replace);
                }
                if (sNewMsg.Length == 0)
                    break;
            }
        }
        return result;
    }

    public bool Add(string sMsg, string sReplaceMsg)
    {
        sMsg = TrimAll(sMsg);
        foreach (var filterMsg in _list)
        {
            if (string.Compare(sMsg, filterMsg.Msg, StringComparison.OrdinalIgnoreCase) == 0)
                return false;
        }
        if (!M2ShareFuncs.CanFilterMsg(sMsg))
            return false;
        _list.Add(new TFilterMsg { Msg = sMsg, Replace = sReplaceMsg });
        _recordCount++;
        return true;
    }

    public bool Delete(string sMsg)
    {
        for (int i = 0; i < _list.Count; i++)
        {
            if (string.Compare(sMsg, _list[i].Msg, StringComparison.OrdinalIgnoreCase) == 0)
            {
                _recordCount--;
                _list.RemoveAt(i);
                return true;
            }
        }
        return false;
    }

    public void LoadFromFile()
    {
        string sFileName = Path.Combine(M2Config.sEnvirDir, "FilterMsgList.txt");
        if (!File.Exists(sFileName))
            return;
        // GBK 一律经 GXX.Core.EncodingInit.GBK 获取：其内部先 Ensure() 注册 CodePagesEncodingProvider，消除加载顺序依赖（CP936 实例等价）。
        foreach (var line in File.ReadLines(sFileName, GXX.Core.EncodingInit.GBK))
        {
            string sLineText = line;
            if (sLineText.Length > 0 && sLineText[0] != ';')
            {
                string sMsg = "", sReplace = "";
                sLineText = TSndaShopList.GetValidStr3(sLineText, ref sMsg, '\t');
                sLineText = TSndaShopList.GetValidStr3(sLineText, ref sReplace, '\t');
                if (M2ShareFuncs.CanFilterMsg(sMsg))
                {
                    _list.Add(new TFilterMsg { Msg = sMsg, Replace = sReplace });
                    _recordCount++;
                }
            }
        }
    }

    public void SaveToFile()
    {
        var saveList = new List<string>();
        foreach (var filterMsg in _list)
            saveList.Add(filterMsg.Msg + "\t" + filterMsg.Replace);
        string sFileName = Path.Combine(M2Config.sEnvirDir, "FilterMsgList.txt");
        try
        {
            File.WriteAllLines(sFileName, saveList, GXX.Core.EncodingInit.GBK);
        }
        catch
        {
            // Delphi try-except 吞异常
        }
    }

    /// <summary>AnsiReplaceText：大小写不敏感的全部替换。</summary>
    internal static string AnsiReplaceText(string text, string from, string to)
    {
        if (from.Length == 0)
            return text;
        var sb = new StringBuilder();
        int pos = 0;
        var cmp = CultureInfo.InvariantCulture.CompareInfo;
        while (pos <= text.Length - from.Length)
        {
            int idx = cmp.IndexOf(text, from, pos, CompareOptions.IgnoreCase);
            if (idx < 0)
                break;
            sb.Append(text, pos, idx - pos);
            sb.Append(to);
            pos = idx + from.Length;
        }
        sb.Append(text, pos, text.Length - pos);
        return sb.ToString();
    }
}
